File: ConvertSortedArrayToBinarySearchTree/ConvertSortedArrayToBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size() == 0)
        	return nullptr;
        TreeNode* root = new TreeNode(0);
        int left = 0;
        int right = nums.size();
        dfs(root, left, right, nums);
        return root;
    }
private:
	void dfs(TreeNode* node, int left, int right, vector<int> &nums) {
		if(left < right) {
			int mid = (left + right) / 2;
			node->val = nums[mid];
			if(left < mid) {
				node->left = new TreeNode(0);
				dfs(node->left, left, mid, nums);
			}
			if(mid + 1 < right) {
				node->right = new TreeNode(0);
				dfs(node->right, mid + 1, right, nums);
			}
		}
		return;
    }
};

File: ConvertSortedArrayToBinarySearchTree/ConvertSortedArrayToBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:
        	return None
        if len(nums) == 1:
        	return TreeNode(nums[0])

        mid = len(nums) / 2
        root = TreeNode(nums[mid])
        left = nums[0:mid]
        right = nums[mid + 1:]
        root.left = self.sortedArrayToBST(left)
        root.right = self.sortedArrayToBST(right)
        return root